### PR TITLE
OSC rdma win allocate: propagate errors to avoid deadlocks

### PR DESCRIPTION
**Pull request by devreal**
This PR changes the shared memory allocation in the rdma osc to synchronize error codes among the ranks in the shared memory communicator to prevent rank 0 from aborting without notifying the other ranks. The changes add two additional allreduce calls on the shared memory communicator and one on the window communicator and replaces two barrier calls with the error code synchronization.

There are still some cases in which ranks abort without notification, most notably if memory allocation fails. I'm not sure what the policy is in Open MPI for this type of error. Catching those would require some additional collective communication, both on the shared memory communicator and the communicator of the window.

Signed-off-by: Joseph Schuchart <schuchart@hlrs.de>

### CODE REVIEW

**Comment by gpaulsen** [comment]
ok to test.

**Comment by hjelmn** [comment]
Well, in my defense I never saw a partial failure :). Looking at this now...

**Comment by hjelmn** [comment]
@devreal Generally we offer no real guarantees when we run out of memory. We do *try* to handles OOM condition gracefully.

File: ompi/mca/osc/rdma/osc_rdma_component.c [hunk 1]
```diff
@@ -593,28 +606,35 @@ static int allocate_state_shared (ompi_osc_rdma_module_t *module, void **base, s
...
         }
 
         module->segment_base = opal_shmem_segment_attach (&module->seg_ds);
         if (NULL == module->segment_base) {
             OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_ERROR, "failed to attach to the shared memory segment");
             ret = OPAL_ERROR;
+        }
+
+        ret = synchronize_errorcode(ret, shared_comm);
+        if (OPAL_UNLIKELY(OMPI_SUCCESS != ret)) {
```
Comment (hjelmn) [inline ompi/mca/osc/rdma/osc_rdma_component.c]: I am curious. Have you run into a scenario where some of the local ranks fail to attach?

File: ompi/mca/osc/rdma/osc_rdma_component.c [hunk 2]
```diff
@@ -1235,6 +1248,9 @@ static int ompi_osc_rdma_component_select (struct ompi_win_t *win, void **base,
 
     /* fill in our part */
     ret = allocate_state_shared (module, base, size);
+
+    /* notify all others if something went wrong */
+    ret = synchronize_errorcode(ret, module->comm);
```
Comment (hjelmn) [inline ompi/mca/osc/rdma/osc_rdma_component.c]: This I see as the most likely scenario. Run into problems on one/some nodes but not all nodes.

File: ompi/mca/osc/rdma/osc_rdma_component.c (hunk 1 above)
Comment (devreal) [reply ompi/mca/osc/rdma/osc_rdma_component.c]: Indeed we have seen this happen when trying to allocate windows beyond 50GB accumulated memory on the node the registration fails (on a ConnectX-3 system). I should have linked the ticket: #6740 

Interestingly, the problem only occurs with the 3.x releases but not on the 4.x releases or master. Still the issue of handling errors exists in all branches.

**Pull request merged**